proposals/native-lock.md

```diff
@@ -0,0 +1,121 @@
+>   - **A warning is reported when `S` is known to be `System.Threading.Lock`.**
+> - From any *class_type* `S` to any *interface_type* `T`, provided `S` implements `T`.
+>   - **A warning is reported when `S` is known to be `System.Threading.Lock`.**
+> - [...]
+
+```cs
+object l = new System.Threading.Lock(); // warning
+lock (l) { } // monitor-based locking is used here
+```
+
+Note that this warning occurs even for equivalent explicit conversions.
+To escape out of the warning and force use of monitor-based locking, one can use
+- the usual warning suppression means (`#pragma warning disable`),
+- `Monitor` APIs directly,
+- indirect casting like `object AsObject<T>(T l) => (object)l;`.
+
+## Alternatives
+[alternatives]: #alternatives
+
+- Support a general pattern that other types can also use to interact with the `lock` keyword.
+  This is a future work that might be implemented when `ref struct`s can participate in generics.
+  Discussed in [LDM 2023-12-04](https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-12-04.md#lock-statement-pattern).
+
+- To avoid ambiguity between the existing monitor-based locking and the new `Lock` (or pattern in the future), we could:
+  - Introduce a new syntax instead of reusing the existing `lock` statement.
+  - Require the new lock types to be `struct`s (since the existing `lock` disallows value types).
+    There could be problems with default constructors and copying if the structs have lazy initialization.
+
+- The codegen could be hardened against thread aborts (which are themselves obsoleted).
+
+- We could warn also when `Lock` is passed as a type parameter, because locking on a type parameter always uses monitor-based locking:
+
+  ```cs
+  M(new Lock()); // could warn here
+
+  void M<T>(T x) // (specifying `where T : Lock` makes no difference)
+  {
+      lock (x) { } // because this uses Monitor
+  }
+  ```
+
+  However, that would cause warnings when storing `Lock`s in a list which is undesirable:
+
+  ```cs
+  List<Lock> list = new();
+  list.Add(new Lock()); // would warn here
+  ```
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Do we want to include static analysis to prevent usage of `System.Threading.Lock` in `using`s with `await`?
+  We could emit either an error or a warning for code like `using (lockVar.EnterLockScope()) { await ... }`.
+  (We would also likely need to consider for this all lock types matching the general pattern if implemented in the future.
+  Although there might need to be an opt-out mechanism as some lock types might be allowed to be used with `await`.)
+  Alternatively, this could be an analyzer shipped as part of the runtime.
+
+- Should the `Lock` type have an `Obsolete` attribute to prevent people with older compilers from accidentally misusing it?
+  (The attribute would be recognized and ignored by compilers supporting the feature.)
```

Comment:
I wonder if the presence of an `Obsolete` attribute could deter people from using the type. Is it common for older compilers to be used to compile against newer libs (like across major .NET versions)?

Reply:
> Is it common for older compilers to be used to compile against newer libs (like across major .NET versions)?

You're right, that doesn't seem common. This `Obsolete` trick is used by the compiler when emitting some types that require new feature to be consumed correctly (e.g., `ref struct`s), but it probably doesn't make sense for types defined in the runtime libs. If someone has a .NET 9 runtime, they will also have .NET 9 compiler (and it's unsupported scenario otherwise anyway).